### PR TITLE
Add tool that calls NHC for all VFNs on chain

### PR DESCRIPTION
**Pull request by banool**
### Description
We will use this tool to periodically check how the VFNs in AIT 3 and beyond are going.

### Test Plan
Run local NHC:
```
cargo run -p aptos-node-checker -- server run --baseline-node-config-paths ~/a/internal-ops/infra/apps/node-checker/configs/ait3_vfn.yaml --listen-address 0.0.0.0
```

Run local testnet:
```
cargo run -p aptos -- node run-local-testnet --with-faucet --faucet-port 8081 --force-restart --assume-yes
```

Run the tool:
```
cargo run -p aptos-vfn-check-client -- --node-address http://127.0.0.1:8080 --nhc-address http://127.0.0.1:20121 --nhc-baseline-config-name ait3_vfn --big-query-key-path ~/a/internal-ops/helm/observability-center/files/bigquery-cron-key.json
```
Output: https://gist.github.com/banool/cf54430db439e216e5216d884d83b44b.

Submitting to BigQuery instead:
```
cargo run -p aptos-vfn-check-client -- --node-address http://127.0.0.1:8080 --nhc-address http://127.0.0.1:20121 --nhc-baseline-config-name ait3_vfn --big-query-key-path ~/a/internal-ops/helm/observability-center/files/bigquery-cron-key.json --output-style big-query
```

- - -
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/3373)

### CODE REVIEW

**Review by rustielin — APPROVED** [review]

File: ecosystem/node-checker/vfn-check-client/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,18 @@
+[package]
```
Comment (JoshLind) [inline ecosystem/node-checker/vfn-check-client/Cargo.toml:1]: nit: I'd update this file to be consistent with the rest of our `Cargo.toml` files 😄 e.g., https://github.com/aptos-labs/aptos-core/blob/c0932968ed1b47c4f8857c78d7008f742356af20/testsuite/forge-cli/Cargo.toml#L1-L0

File: ecosystem/node-checker/vfn-check-client/README.md [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+2. Process that to create a map from operator account address to VFN network addresses.
+3. Query NHC for each VFN.
+4. Push the results to BigQuery.
+
+The original intent behind this tool is to confirm operators are running quality, operational VFNs as part of AIT3. This tool can be easily adapted for other use cases down the line.
+
+## Local development
+Run a local network with both a validator and VFNs:
+```
+cargo run -p forge-cli -- --suite "state_sync" --num-validators 1 --num-validator-fullnodes 2 --mempool-backlog 5000 test local-swarm
```
Comment (JoshLind) [inline ecosystem/node-checker/vfn-check-client/README.md:15]: nit: I would use the "run_forever" suite so the cluster doesn't die:

```
cargo run -p forge-cli -- --suite "run_forever" --num-validators 2 --num-validator-fullnodes 2 test local-swarm
```

Comment (JoshLind) [reply ecosystem/node-checker/vfn-check-client/README.md:15]: Also, `num-validators` needs to be bigger than `num-validator-fullnodes`

File: ecosystem/node-checker/vfn-check-client/src/big_query.rs [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+    .await;
+
+    info!(
+        "Creating dataset if necessary: {}",
+        big_query_args.big_query_dataset_id
+    );
+
+    // Create the dataset if necessary.
+    ignore_already_exists_error(
+        client
```
Comment (JoshLind) [inline ecosystem/node-checker/vfn-check-client/src/big_query.rs:112]: nit: I'd maybe move each of the dataset creation calls into helpers 😄, e.g., the one below is pretty big.

File: ecosystem/node-checker/vfn-check-client/src/check.rs [hunk 4]
```diff
@@ -0,0 +1,344 @@
...
+use tokio::sync::Semaphore;
+
+use crate::helpers::extract_network_address;
+
+// Unfortunately we don't have any way, on chain or not, to know what the API
+// port is, so we just assume it is 80 or 8080. If their VFN is inaccessible at 80,
+// e.g. because it runs on 7777 and they have an LB in front of it that listens
+// at 80, we're just out of luck. If we get this kind of information from
+// elsewhere at some point, we could look at joining it here.
+pub const API_PORTS: &[u16] = &[80, 8080];
```
Comment (JoshLind) [inline ecosystem/node-checker/vfn-check-client/src/check.rs:32]: Annoying... In a non code-freeze world, we could have exposed this using the REST API 😄

File: ecosystem/node-checker/vfn-check-client/src/helpers.rs [hunk 5]
```diff
@@ -0,0 +1,24 @@
+// Copyright (c) Aptos
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::{Context, Result};
+use aptos_sdk::types::network_address::NetworkAddress;
+use std::net::{SocketAddr, ToSocketAddrs};
+
+// This function takes a NetworkAddress and returns a string representation
+// of it if it is a format we can send to NHC. Otherwise we return an error.
+pub fn extract_network_address(network_address: &NetworkAddress) -> Result<String> {
```
Comment (JoshLind) [inline ecosystem/node-checker/vfn-check-client/src/helpers.rs:10]: nit: why it's own file if there's only one caller? 😄

**Review by JoshLind — APPROVED** [review]
Thanks @banool. Looks reasonable to me. Mostly just nits.

File: ecosystem/node-checker/vfn-check-client/src/helpers.rs (hunk 5 above)
Comment (banool) [reply ecosystem/node-checker/vfn-check-client/src/helpers.rs:10]: Smaller files hahah but fair point, I can merge / move more stuff here like you suggested.

File: ecosystem/node-checker/vfn-check-client/src/check.rs (hunk 4 above)
Comment (banool) [reply ecosystem/node-checker/vfn-check-client/src/check.rs:32]: Alas, we don't know what the REST API is, it's not on chain nor given to us any other way :p

**Pull request merged**